### ISSUE TITLE
你好，请问STViT论文中Intra and inter-window spatial pooling. 这部分代码在哪里？我没有找到关于生成global tokens的代码。

### ISSUE THREAD

**Issue by 9527950** [changsn/STViT-R#4]

**Comment by 9527950** [comment]
![微信图片_20240606114512](https://github.com/changsn/STViT-R/assets/82935515/a0a1e584-6e46-4a9c-b1dd-c7f325405c49)

**Comment by changsn** [comment]
Send the code to you via email.

**Comment by laplace522** [comment]
> Send the code to you via email.

你好，请问您可以给我这段代码吗

**Comment by changsn** [comment]
> > Send the code to you via email.
> 
> 你好，请问您可以给我这段代码吗
```
class weighted_pooling(nn.Module):
    num_samples = 16
    def __init__(self, dim):
        super().__init__()
        self.gen_mask = nn.Sequential(
            nn.Conv2d(dim, dim, 3, stride=1, padding=1, groups=dim),
            LayerNormProxy(dim),
            nn.GELU(),
            nn.Conv2d(dim, 1, 1, 1, 0, bias=False)
        )

    def forward(self, x):
        B, C, H, W = x.size()
        zero_pad_mask = torch.zeros(B, 1, H, W, device=x.device)
        zero_pad_mask = F.pad(zero_pad_mask, (1, 1, 1, 1), value=-100)
        x = F.unfold(x, kernel_size=4, padding=1, stride=4).reshape(B, C, 4, 4, 16).permute(0, 4, 1, 2, 3).reshape(-1, C, 4, 4)
        zero_pad_mask = F.unfold(zero_pad_mask, kernel_size=4, padding=0, stride=4).reshape(B, 1, 4, 4, 16).permute(0, 4, 1, 2, 3).reshape(-1, 1, 4, 4)
        mask = self.gen_mask(x)  # B*16, 1, 4, 4
        mask += zero_pad_mask
        mask = torch.softmax(mask.reshape(-1, 1, 16), dim=-1)
        weighted_x = torch.sum(x.reshape(-1, C,  16) * mask, dim=-1)  # (B*16, C)
        weighted_x = weighted_x.reshape(B, 16, C)
        return weighted_x


class weighted_offset_pooling(nn.Module):
    num_samples = 100
    def __init__(self, dim):
        super().__init__()
        self.k = 2
        self.l = 10
        self.t = self.l * self.l
        self.p = 3
        self.gen_mask = nn.Sequential(
            nn.Conv2d(dim, dim, 3, stride=1, padding=1, groups=dim),
            LayerNormProxy(dim),
            nn.GELU(),
            nn.Conv2d(dim, 1, 1, 1, 0, bias=False)
        )
        self.offset = nn.Sequential(
            nn.Conv2d(dim, dim, 3, stride=1, padding=1, groups=dim),
            LayerNormProxy(dim),
            nn.GELU(),
            nn.Conv2d(dim, self.k*self.k, 1, 1, 0, bias=False)
        )

    def forward(self, x):
        B, C, H, W = x.size()
        zero_pad_mask = torch.zeros(B, 1, H, W, device=x.device)
        zero_pad_mask = F.pad(zero_pad_mask, (self.p, self.p, self.p, self.p), value=-100)
        x = F.unfold(x, kernel_size=self.k, padding=self.p, stride=self.k).reshape(B, C, self.k, self.k, self.t).permute(0, 4, 1, 2, 3).reshape(-1, C, self.k, self.k)
        zero_pad_mask = F.unfold(zero_pad_mask, kernel_size=self.k, padding=0, stride=self.k).reshape(B, 1, self.k, self.k, self.t).permute(0, 4, 1, 2, 3).reshape(-1, 1, self.k, self.k)
        mask = self.gen_mask(x)  # B*16, 1, 4, 4
        mask += zero_pad_mask
        # mask = torch.softmax(mask.reshape(-1, 1, 16), dim=-1)
        weighted_x = torch.sum(x.reshape(-1, C, self.k*self.k) * torch.softmax(mask.reshape(-1, 1, self.k*self.k), dim=-1), dim=-1)  # (B*16, C)
        # weighted_x = weighted_x.reshape(B, 16, C)
        weighted_x = weighted_x.reshape(B, self.l, self.l, C).permute(0, 3, 1, 2)
        # pdb.set_trace()
        offset = self.offset(weighted_x).reshape(B, self.k, self.k, self.t).permute(0, 3, 1, 2).reshape(B*self.t, 1, self.k, self.k)  # B, 16, 4, 4
        mask = mask + offset
        weighted_x = torch.sum(x.reshape(-1, C, self.k*self.k) * torch.softmax(mask.reshape(-1, 1, self.k*self.k), dim=-1), dim=-1)
        weighted_x = weighted_x.reshape(B, self.t, C)
        return weighted_x
```

**Comment by laplace522** [comment]
非常感谢您发来的代码，我已经收到了。




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "changsn/STViT-R"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年7月8日(星期一) 下午4:20
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [changsn/STViT-R] 你好，请问STViT论文中Intra and inter-window spatial pooling. 这部分代码在哪里？我没有找到关于生成global tokens的代码。 (Issue #4)





   
Send the code to you via email.
  
你好，请问您可以给我这段代码吗
  
class weighted_pooling(nn.Module):
 num_samples = 16
 def init(self, dim):
 super().init()
 self.gen_mask = nn.Sequential(
 nn.Conv2d(dim, dim, 3, stride=1, padding=1, groups=dim),
 LayerNormProxy(dim),
 nn.GELU(),
 nn.Conv2d(dim, 1, 1, 1, 0, bias=False)
 )
 def forward(self, x):     B, C, H, W = x.size()     zero_pad_mask = torch.zeros(B, 1, H, W, device=x.device)     zero_pad_mask = F.pad(zero_pad_mask, (1, 1, 1, 1), value=-100)     x = F.unfold(x, kernel_size=4, padding=1, stride=4).reshape(B, C, 4, 4, 16).permute(0, 4, 1, 2, 3).reshape(-1, C, 4, 4)     zero_pad_mask = F.unfold(zero_pad_mask, kernel_size=4, padding=0, stride=4).reshape(B, 1, 4, 4, 16).permute(0, 4, 1, 2, 3).reshape(-1, 1, 4, 4)     mask = self.gen_mask(x)  # B*16, 1, 4, 4     mask += zero_pad_mask     mask = torch.softmax(mask.reshape(-1, 1, 16), dim=-1)     weighted_x = torch.sum(x.reshape(-1, C,  16) * mask, dim=-1)  # (B*16, C)     weighted_x = weighted_x.reshape(B, 16, C)     return weighted_x  
class weighted_offset_pooling(nn.Module):
 num_samples = 100
 def init(self, dim):
 super().init()
 self.k = 2
 self.l = 10
 self.t = self.l * self.l
 self.p = 3
 self.gen_mask = nn.Sequential(
 nn.Conv2d(dim, dim, 3, stride=1, padding=1, groups=dim),
 LayerNormProxy(dim),
 nn.GELU(),
 nn.Conv2d(dim, 1, 1, 1, 0, bias=False)
 )
 self.offset = nn.Sequential(
 nn.Conv2d(dim, dim, 3, stride=1, padding=1, groups=dim),
 LayerNormProxy(dim),
 nn.GELU(),
 nn.Conv2d(dim, self.k*self.k, 1, 1, 0, bias=False)
 )
 def forward(self, x):     B, C, H, W = x.size()     zero_pad_mask = torch.zeros(B, 1, H, W, device=x.device)     zero_pad_mask = F.pad(zero_pad_mask, (self.p, self.p, self.p, self.p), value=-100)     x = F.unfold(x, kernel_size=self.k, padding=self.p, stride=self.k).reshape(B, C, self.k, self.k, self.t).permute(0, 4, 1, 2, 3).reshape(-1, C, self.k, self.k)     zero_pad_mask = F.unfold(zero_pad_mask, kernel_size=self.k, padding=0, stride=self.k).reshape(B, 1, self.k, self.k, self.t).permute(0, 4, 1, 2, 3).reshape(-1, 1, self.k, self.k)     mask = self.gen_mask(x)  # B*16, 1, 4, 4     mask += zero_pad_mask     # mask = torch.softmax(mask.reshape(-1, 1, 16), dim=-1)     weighted_x = torch.sum(x.reshape(-1, C, self.k*self.k) * torch.softmax(mask.reshape(-1, 1, self.k*self.k), dim=-1), dim=-1)  # (B*16, C)     # weighted_x = weighted_x.reshape(B, 16, C)     weighted_x = weighted_x.reshape(B, self.l, self.l, C).permute(0, 3, 1, 2)     # pdb.set_trace()     offset = self.offset(weighted_x).reshape(B, self.k, self.k, self.t).permute(0, 3, 1, 2).reshape(B*self.t, 1, self.k, self.k)  # B, 16, 4, 4     mask = mask + offset     weighted_x = torch.sum(x.reshape(-1, C, self.k*self.k) * torch.softmax(mask.reshape(-1, 1, self.k*self.k), dim=-1), dim=-1)     weighted_x = weighted_x.reshape(B, self.t, C)     return weighted_x
